### PR TITLE
Docs for Replication and target portal IPs for Primera/3PAR CSP

### PR DESCRIPTION
**Pull request by sneharai4**

### CODE REVIEW

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 1]
```diff
@@ -194,6 +197,45 @@ In the HPE Primera or 3PAR Storage system, the QoS rules are applied to a volume
...
+  - targetCpg: FC_r6
+    targetSnapCpg: FC_r6
+    targetName: CSSOS-SSA05
+    targetSecret: secret-r
+    targetSecretNamespace: kube-system
+```
+
+!!! important
+    • targetCpg, targetName, targetSecret and targetSecretNamespace are mandatory for `HPEReplicationDeviceInfo` CRD.<br />
+    • Replication mode is set to sync for peer persistence by default.<br />
```
Comment (c-snell) [inline docs/container_storage_provider/hpe_3par_primera/index.md:226]:
```suggestion
    • Currently, the HPE CSI Driver only supports Remote Copy Peer Persistence mode. Async support will be added in a future release.<br />
```

This needs to be moved the beginning of the replication section and highlighted.

Currently, the HPE CSI Driver only supports Remote Copy Peer Persistence mode. Async support will be added in a future release.

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 2]
```diff
@@ -194,6 +197,45 @@ In the HPE Primera or 3PAR Storage system, the QoS rules are applied to a volume
...
+    • Create this custom resource object before creating storage class referring this parameter.<br />
```
Comment (c-snell) [inline docs/container_storage_provider/hpe_3par_primera/index.md:227]: Remove due to being redundant information

```suggestion
```

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 3]
```diff
@@ -194,6 +197,45 @@ In the HPE Primera or 3PAR Storage system, the QoS rules are applied to a volume
...
 
+### Peer Persistence
+
+These parameters are applicable only for replication. Both parameters are mandatory. Remote copy group name given in SC if not present will be created.
+
+| Parameter          | Option  | Description |
+| ------------------ | ------- | ----------- |
+| remoteCopyGroup    | Text    | Name of a new or existing remote copy group on Primera/3PAR array. |
+| replicationDevices | Text    | Indicates name of custom resource of type `hpereplicationdeviceinfos`. |
+
```
Comment (c-snell) [inline docs/container_storage_provider/hpe_3par_primera/index.md:208]: Section needed.

To enable replication within the HPE CSI Driver, the following steps must be completed:
* Create `Secrets` for both primary and target HPE Primera or 3PAR arrays. Refer to [Adding additional backends](https://scod.hpedev.io/csi_driver/deployment.html#adding_additional_backends)
* Create replication custom resource 
* Create replication enabled `StorageClass`

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 4]
```diff
@@ -194,6 +197,45 @@ In the HPE Primera or 3PAR Storage system, the QoS rules are applied to a volume
 | ------------------ | ------- | ----------- |
 | qosName      | Text         | Name of the HPE Primera or 3PAR volume set which has QoS rules. This parameter is optional. If specified, the `PersistentVolumeClaim` will be associated with the HPE Primera or 3PAR volume set, for purposes of applying the QoS rules. |
 
+### Peer Persistence
```
Comment (c-snell) [inline docs/container_storage_provider/hpe_3par_primera/index.md:200]: Keep with official name: Remote Copy with Peer Persistence synchronous replication
```suggestion
### Remote Copy with Peer Persistence synchronous replication
```

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 5]
```diff
@@ -194,6 +197,45 @@ In the HPE Primera or 3PAR Storage system, the QoS rules are applied to a volume
...
+
+These parameters are applicable only for replication. Both parameters are mandatory. Remote copy group name given in SC if not present will be created.
```
Comment (c-snell) [inline docs/container_storage_provider/hpe_3par_primera/index.md:202]: 1. Move the parameters section below to follow order of operation to enable replication.

2. Move this section here to outline the steps that will be followed. 
```
To enable replication within the HPE CSI Driver, the following steps must be completed:

* Create Secrets for both primary and target HPE Primera or 3PAR arrays. Refer to Adding additional backends
* Create replication custom resource
* Create replication enabled StorageClass

For a tutorial on how to enable replication, check out the blog [Enabling Remote Copy using the HPE CSI Driver for Kubernetes on HPE Primera](https://developer.hpe.com/blog/ppPAlQ807Ah8QGMNl1YE/tutorial-enabling-remote-copy-using-the-hpe-csi-driver-for-kubernetes-on)
```

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 6]
```diff
@@ -194,6 +197,45 @@ In the HPE Primera or 3PAR Storage system, the QoS rules are applied to a volume
...
+
+### Target Portal IPs
+
+This parameter is applicable when user wants to choose specific iSCSI target portal IP on which iSCSI sessions needs to be established.
+
+| Parameter      | Option  | Description |
+| -------------- | ------- | ----------- |
+| iscsiPortalIps | Text | Comma separated list of target portal IPs. |
+
+!!! note
```
Comment (c-snell) [inline docs/container_storage_provider/hpe_3par_primera/index.md:237]: Remove as inserted above.

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 7]
```diff
@@ -194,6 +197,45 @@ In the HPE Primera or 3PAR Storage system, the QoS rules are applied to a volume
...
+```
+
+!!! important
+    • targetCpg, targetName, targetSecret and targetSecretNamespace are mandatory for `HPEReplicationDeviceInfo` CRD.<br />
+    • Replication mode is set to sync for peer persistence by default.<br />
+    • Create this custom resource object before creating storage class referring this parameter.<br />
+
+### Target Portal IPs
+
+This parameter is applicable when user wants to choose specific iSCSI target portal IP on which iSCSI sessions needs to be established.
```
Comment (c-snell) [inline docs/container_storage_provider/hpe_3par_primera/index.md:231]:
```suggestion
This parameter allows the ability to specify a subset of HPE Primera/3PAR iSCSI ports for iSCSI sessions. By default, the HPE CSI Driver uses all available iSCSI ports.
```

**Review by c-snell — CHANGES_REQUESTED** [review]
Please review suggested changes.

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 8]
```diff
@@ -89,6 +89,9 @@ These parameters are used for volume provisioning and supported platforms.
 | cloneOf  | Text      | Name of the `PersistentVolumeClaim` to clone. | **X** | **X** |
 | virtualCopyOf  | Text      | Name of the `PersistentVolumeClaim` to snapshot. | **X** | **X** |
 | qosName  | Text      | Name of the volume set which has QoS rules applied. | **X** | **X** |
+| remoteCopyGroup <br /> | Text | Name of a new or existing remote copy group on Primera/3PAR array. | **X** | **X** |
```
[the remark this answers is not in the document]
Comment (sneharai4) [reply docs/container_storage_provider/hpe_3par_primera/index.md:92]: Done.

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 9]
```diff
@@ -89,6 +89,9 @@ These parameters are used for volume provisioning and supported platforms.
...
+| replicationDevices <br /> | Text <br /> | Indicates name of custom resource of type `hpereplicationdeviceinfos`. | **X** | **X** |
+| iscsiPortalIps <br /> | Text <br /> | Comma separated list of target portal IPs. | **X** | **X** |
```
[the remark this answers is not in the document]
Comment (sneharai4) [reply docs/container_storage_provider/hpe_3par_primera/index.md:94]: Done.

File: docs/container_storage_provider/hpe_3par_primera/index.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (sneharai4) [reply docs/container_storage_provider/hpe_3par_primera/index.md:202]: Done.

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 10]
```diff
@@ -194,6 +197,45 @@ In the HPE Primera or 3PAR Storage system, the QoS rules are applied to a volume
 | ------------------ | ------- | ----------- |
 | qosName      | Text         | Name of the HPE Primera or 3PAR volume set which has QoS rules. This parameter is optional. If specified, the `PersistentVolumeClaim` will be associated with the HPE Primera or 3PAR volume set, for purposes of applying the QoS rules. |
 
+### Peer Persistence
+
+These parameters are applicable only for replication. Both parameters are mandatory. Remote copy group name given in SC if not present will be created.
+
+| Parameter          | Option  | Description |
+| ------------------ | ------- | ----------- |
+| remoteCopyGroup    | Text    | Name of a new or existing remote copy group on Primera/3PAR array. |
```
[the remark this answers is not in the document]
Comment (sneharai4) [reply docs/container_storage_provider/hpe_3par_primera/index.md:206]: Done.

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 11]
```diff
@@ -194,6 +197,45 @@ In the HPE Primera or 3PAR Storage system, the QoS rules are applied to a volume
...
+| replicationDevices | Text    | Indicates name of custom resource of type `hpereplicationdeviceinfos`. |
```
[the remark this answers is not in the document]
Comment (sneharai4) [reply docs/container_storage_provider/hpe_3par_primera/index.md:207]: Done.

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 12]
```diff
@@ -194,6 +197,45 @@ In the HPE Primera or 3PAR Storage system, the QoS rules are applied to a volume
...
+
+A Custom Resource Definition (CRD) named `hpereplicationdeviceinfos.storage.hpe.com` holds target array details. This object name needs to be used in SC for `replicationDevices`.
```
[the remark this answers is not in the document]
Comment (sneharai4) [reply docs/container_storage_provider/hpe_3par_primera/index.md:209]: Done.

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 13]
```diff
@@ -194,6 +197,45 @@ In the HPE Primera or 3PAR Storage system, the QoS rules are applied to a volume
...
+```yaml
+apiVersion: storage.hpe.com/v1
+kind: HPEReplicationDeviceInfo
+metadata:
+  name: r1
+spec:
+  target_array_details:
+  - targetCpg: FC_r6
```
[the remark this answers is not in the document]
Comment (sneharai4) [reply docs/container_storage_provider/hpe_3par_primera/index.md:217]: Done.

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 14]
```diff
@@ -194,6 +197,45 @@ In the HPE Primera or 3PAR Storage system, the QoS rules are applied to a volume
...
+    targetSnapCpg: FC_r6
```
[the remark this answers is not in the document]
Comment (sneharai4) [reply docs/container_storage_provider/hpe_3par_primera/index.md:218]: Done.

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 15]
```diff
@@ -194,6 +197,45 @@ In the HPE Primera or 3PAR Storage system, the QoS rules are applied to a volume
...
+    targetName: CSSOS-SSA05
```
[the remark this answers is not in the document]
Comment (sneharai4) [reply docs/container_storage_provider/hpe_3par_primera/index.md:219]: Done.

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 16]
```diff
@@ -194,6 +197,45 @@ In the HPE Primera or 3PAR Storage system, the QoS rules are applied to a volume
...
+    targetSecret: secret-r
```
[the remark this answers is not in the document]
Comment (sneharai4) [reply docs/container_storage_provider/hpe_3par_primera/index.md:220]: Done.

File: docs/container_storage_provider/hpe_3par_primera/index.md (hunk 2 above)
Comment (sneharai4) [reply docs/container_storage_provider/hpe_3par_primera/index.md:227]: Done.

File: docs/container_storage_provider/hpe_3par_primera/index.md (hunk 3 above)
Comment (sneharai4) [reply docs/container_storage_provider/hpe_3par_primera/index.md:208]: Done.

File: docs/container_storage_provider/hpe_3par_primera/index.md (hunk 4 above)
Comment (sneharai4) [reply docs/container_storage_provider/hpe_3par_primera/index.md:200]: Done.

File: docs/container_storage_provider/hpe_3par_primera/index.md (hunk 7 above)
Comment (sneharai4) [reply docs/container_storage_provider/hpe_3par_primera/index.md:231]: Done.

File: docs/container_storage_provider/hpe_3par_primera/index.md (hunk 6 above)
Comment (sneharai4) [reply docs/container_storage_provider/hpe_3par_primera/index.md:237]: Done.

File: docs/container_storage_provider/hpe_3par_primera/index.md (hunk 5 above)
Comment (sneharai4) [reply docs/container_storage_provider/hpe_3par_primera/index.md:202]: Done.

File: docs/container_storage_provider/hpe_3par_primera/index.md (hunk 1 above)
Comment (sneharai4) [reply docs/container_storage_provider/hpe_3par_primera/index.md:226]: Done.

**Comment by sneharai4** [comment]
@c-snell https://developer.hpe.com/blog/ppPAlQ807Ah8QGMNl1YE/tutorial-enabling-remote-copy-using-the-hpe-csi-driver-for-kubernetes-on link is not available.
 It gives a 'Page not found...'

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 17]
```diff
@@ -194,6 +197,49 @@ In the HPE Primera or 3PAR Storage system, the QoS rules are applied to a volume
...
+
+For a tutorial on how to enable replication, check out the blog [Enabling Remote Copy using the HPE CSI Driver for Kubernetes on HPE Primera](https://developer.hpe.com/blog/ppPAlQ807Ah8QGMNl1YE/tutorial-enabling-remote-copy-using-the-hpe-csi-driver-for-kubernetes-on)
+
+These parameters are applicable only for replication. Both parameters are mandatory. If the remote copy volume group (RCG) name, as defined within the `StorageClass`, does not exist on the HPE Primera or 3PAR array, then a new RCG will be created.
+
+| Parameter          | Option  | Description |
+| ------------------ | ------- | ----------- |
+| remoteCopyGroup    | Text    | Name of new or existing remote copy group on the HPE Primera/3PAR array. |
+| replicationDevices | Text    | Indicates name of `hpereplicationdeviceinfos` Custom Resource Definition (CRD). |
+
```
Comment (c-snell) [inline docs/container_storage_provider/hpe_3par_primera/index.md:216]: Move below CRD example.

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 18]
```diff
@@ -194,6 +197,49 @@ In the HPE Primera or 3PAR Storage system, the QoS rules are applied to a volume
...
 
+### Remote Copy with Peer Persistence synchronous replication
+
+To enable replication within the HPE CSI Driver, the following steps must be completed:
+
+* Create `Secrets` for both primary and target HPE Primera or 3PAR arrays. Refer to [Adding additional backends](../../csi_driver/deployment.md#adding_additional_backends).
+* Create replication custom resource.
+* Create replication enabled `StorageClass`.
+
+For a tutorial on how to enable replication, check out the blog [Enabling Remote Copy using the HPE CSI Driver for Kubernetes on HPE Primera](https://developer.hpe.com/blog/ppPAlQ807Ah8QGMNl1YE/tutorial-enabling-remote-copy-using-the-hpe-csi-driver-for-kubernetes-on)
```
Comment (c-snell) [inline docs/container_storage_provider/hpe_3par_primera/index.md:208]: This link is to the soon to be release peer persistence blog. The link is valid and will go live on GA date (9/14).

**Review by c-snell — CHANGES_REQUESTED** [review]

File: docs/container_storage_provider/hpe_3par_primera/index.md (hunk 17 above)
Comment (sneharai4) [reply docs/container_storage_provider/hpe_3par_primera/index.md:216]: Done.

File: docs/container_storage_provider/hpe_3par_primera/index.md (hunk 18 above)
Comment (sneharai4) [reply docs/container_storage_provider/hpe_3par_primera/index.md:208]: Okay.

**Review by c-snell — APPROVED** [review]
LGTM

**Pull request merged**